src/setup.rs

```diff
@@ -177,30 +225,89 @@ macro_rules! impl_pins {
+    feature = "device-selected",
+    any(feature = "stm32f4xx-hal", feature = "stm32f7xx-hal")
+))]
 impl_pins!(
     RmiiRefClk: [
-        PA1<Input<Floating>>,
+        PA1<Input>,
     ],
     RmiiCrsDv: [
-        PA7<Input<Floating>>,
+        PA7<Input>,
     ],
     RmiiTxEN: [
-        PB11<Input<Floating>>,
-        PG11<Input<Floating>>,
+        PB11<Input>,
+        PG11<Input>,
     ],
     RmiiTxD0: [
-        PB12<Input<Floating>>,
-        PG13<Input<Floating>>,
+        PB12<Input>,
+        PG13<Input>,
     ],
     RmiiTxD1: [
-        PB13<Input<Floating>>,
-        PG14<Input<Floating>>,
+        PB13<Input>,
+        PG14<Input>,
     ],
     RmiiRxD0: [
-        PC4<Input<Floating>>,
+        PC4<Input>,
     ],
     RmiiRxD1: [
-        PC5<Input<Floating>>,
+        PC5<Input>,
     ],
 );
+
+#[cfg(feature = "stm32f1xx-hal")]
+mod stm32f1 {
+    use super::*;
+    use stm32f1xx_hal::gpio::{
+        gpioa, gpiob, gpioc, gpiod, Alternate, Floating, IOPinSpeed, Input, OutputSpeed, PushPull,
+    };
+
+    // STM32F1xx's require access to the CRL/CRH registers to change pin mode. As a result, we
+    // require that pins are already in the necessary mode before constructing `EthPins` as it
+    // would be inconvenient to pass CRL and CRH through to the `AlternateVeryHighSpeed` callsite.
+
+    type PA1 = gpioa::PA1<Input<Floating>>;
+    type PA2 = gpioa::PA2<Alternate<PushPull>>;
+    type PA7 = gpioa::PA7<Input<Floating>>;
+    type PB11 = gpiob::PB11<Alternate<PushPull>>;
+    type PB12 = gpiob::PB12<Alternate<PushPull>>;
+    type PB13 = gpiob::PB13<Alternate<PushPull>>;
+    type PC1 = gpioc::PC1<Alternate<PushPull>>;
+    type PC4 = gpioc::PC4<Input<Floating>>;
+    type PC5 = gpioc::PC5<Input<Floating>>;
+    type PD8 = gpiod::PD8<Input<Floating>>;
+    type PD9 = gpiod::PD9<Input<Floating>>;
+    type PD10 = gpiod::PD10<Input<Floating>>;
+
+    unsafe impl RmiiRefClk for PA1 {}
+    unsafe impl RmiiCrsDv for PA7 {}
+    unsafe impl RmiiCrsDv for PD8 {}
+    unsafe impl RmiiTxEN for PB11 {}
+    unsafe impl RmiiTxD0 for PB12 {}
+    unsafe impl RmiiTxD1 for PB13 {}
+    unsafe impl RmiiRxD0 for PC4 {}
+    unsafe impl RmiiRxD0 for PD9 {}
+    unsafe impl RmiiRxD1 for PC5 {}
+    unsafe impl RmiiRxD1 for PD10 {}
+
+    macro_rules! impl_alt_very_high_speed {
+        ($($PIN:ident),*) => {
+            $(
+                impl AlternateVeryHighSpeed for $PIN {
+                    fn into_af11_very_high_speed(self) {
+                        // SAFETY: this assumes that no other access to the GPIO
+                        // control registers is being performed at the same time.
+                        //
+                        // With the current API, this is the best we can do.
+                        let acrl: &mut _ = &mut unsafe { core::mem::transmute(()) };
+                        let mut pin = self.into_alternate_push_pull(acrl);
+                        pin.set_speed(acrl, IOPinSpeed::Mhz50);
```

Comment:
This is clever indeed. However, it's still unsound and a bit hard to make it sound.

We could use `cortex_m::interrupt::free` to make your assumption true, but that still leaves us with the `transmute`, which is sound right now (with `interrupt::free`), but the HAL can make a simple semver compatible change and make this unsound and we would be the ones to blame.

I see two alternatives:
1. Just go back to the documentation stating that the pin should be set to high speed, which is quite fine, given that it's the default behavior for the HAL today.
2. Use `interrupt::free` and change the registers manually ourselves (with the pac). This is quite cumbersome due to the several different pins involved.

I would be fine with either one.

Reply:
Thanks for the look-over :)

I've added a critical section around the use of the `unsafe`ly construced `CRL` register now. I've also added a comment explaining why it should be impossible for a change to the f1 HAL to introduce unsound behaviour.

Reply:
Right, this is good enough, we already use critical sections in other configurations, so we're already risking some problems if this initial configuration is done in an interrupt, which would be quite weird anyways.